tfx_addons/xgboost_evaluator/xgboost_predict_extractor.py

```diff
@@ -0,0 +1,208 @@
+  def __init__(self, eval_shared_models: Dict[Text, types.EvalSharedModel],
+               eval_config: tfma.EvalConfig):
+    super().__init__(
+        {k: v.model_loader
+         for k, v in eval_shared_models.items()})
+    self._eval_config = eval_config
+
+  def setup(self):
+    """Function that accesses and saves the chosen feature keys and label key."""
+
+    # models are loaded and stored into self._loaded_models below
+    super().setup()
+    self._feature_keys = None
+    self._label_key = None
+
+    # check to see whether each of these loaded models has a corresponding model spec
+    if self._eval_config:
+      label_config = self.extract_model_specs()
+      for name in self._loaded_models:
+        if name not in label_config:
+          raise ValueError(f"Missing model spec for loaded model {name}.")
+
+    for name, loaded_model in self._loaded_models.items():
+      feature_keys = loaded_model.feature_names
+      if self._feature_keys and self._label_key:
+        assert self._feature_keys == feature_keys, (
+            f'Features mismatch in loaded models. Expected {self._feature_keys}'
+            f', got {lfeature_keys} instead.')
+        assert self._label_key == label_config[name], (
+            f'Label mismatch in loaded models. Expected "{self._label_key}"'
+            f', got "{label_config[name]}" instead.')
+      elif feature_keys and label_config[name]:
+        self._feature_keys = feature_keys
+        self._label_key = label_config[name]
+      else:
+        raise ValueError(
+            f'Missing feature or label keys in loaded model {name}.')
+
+  def extract_model_specs(self):
+    label_specs = {}
+    for config in self._eval_config.model_specs:
+      if config.name:
+        label_specs[config.name] = config.label_key
+      else:  # if the input name to ModelSpec is None, ModelSpec doesn't save it and config.name resolves to ''.
+        label_specs[None] = config.label_key
+    return label_specs
+
+  def process(self, elem: types.Extracts) -> Iterable[types.Extracts]:
+    """Uses loaded models to make predictions on batches of data.
+
+    Args:
+      elem: An extract containing batched features.
+
+    Yields:
+      Copy of the original extracts with predictions added for each model. If
+      there are multiple models, a list of dicts keyed on model names will be
+      added, with each value corresponding to a prediction for a single sample.
+    """
+    # Build feature and label vectors because xgboost cannot read tf.Examples.
+    features = []
+    labels = []
+    result = copy.copy(elem)
+    for features_dict in result[constants.FEATURES_KEY]:
+      features_row = [features_dict[key] for key in self._feature_keys]
+      features.append(np.concatenate(features_row))
+      labels.append(features_dict[self._label_key])
+    result[constants.LABELS_KEY] = np.concatenate(labels)
+    features = xgb.DMatrix(pd.DataFrame(features, columns=self._feature_keys))
```

Comment:
do we need to load this into a dataframe before passing it into DMatrix?

Reply:
I don't believe so - the sklearn example does something similar to this in its example, and that works fine.